x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.ts

```diff
@@ -0,0 +1,271 @@
+        setTokenReadWrite: (activeApiToken, { name, checked }) => ({
+          ...activeApiToken,
+          [name]: checked,
+        }),
+        setTokenType: (activeApiToken, tokenType) => ({
+          ...activeApiToken,
+          access_all_engines: tokenType === ADMIN ? false : activeApiToken.access_all_engines,
+          engines: tokenType === ADMIN ? [] : activeApiToken.engines,
+          write: tokenType === PRIVATE,
+          read: tokenType === PRIVATE,
+          type: tokenType,
+        }),
+        toggleCredentialsForm: (_, activeApiToken) => activeApiToken || defaultApiToken,
+      },
+    ],
+    activeApiTokenRawName: [
+      '',
+      {
+        setTokenName: (_, activeApiTokenRawName) => activeApiTokenRawName,
+        toggleCredentialsForm: (activeApiTokenRawName, activeApiToken) =>
+          activeApiToken.name || activeApiTokenRawName,
+        hideCredentialsForm: () => '',
+        onApiTokenCreateSuccess: () => '',
+        onApiTokenUpdateSuccess: () => '',
+      },
+    ],
+    activeApiTokenIsExisting: [
+      false,
+      {
+        toggleCredentialsForm: (_, activeApiToken) => !!activeApiToken.id,
+      },
+    ],
+    showCredentialsForm: [
+      false,
+      {
+        toggleCredentialsForm: (showCredentialsForm) => !showCredentialsForm,
+        hideCredentialsForm: () => false,
+        onApiTokenCreateSuccess: () => false,
+        onApiTokenUpdateSuccess: () => false,
+      },
+    ],
+    formErrors: [
+      [],
+      {
+        onApiTokenError: (_, formErrors) => formErrors,
+        onApiTokenCreateSuccess: () => [],
+        toggleCredentialsForm: () => [],
+        resetCredentials: () => [],
+      },
+    ],
+    // TODO flashMessages from ent-search
+  }),
+  selectors: ({ selectors }) => ({
+    // TODO fullEngineAccessChecked from ent-search
+    dataLoading: [
+      () => [selectors.isCredentialsDetailsComplete, selectors.isCredentialsDataComplete],
+      (isCredentialsDetailsComplete, isCredentialsDataComplete) => {
+        return isCredentialsDetailsComplete === false || isCredentialsDataComplete === false;
+      },
+    ],
+    keyHelpText: [
+      () => [selectors.activeApiToken, selectors.activeApiTokenRawName],
+      (activeApiToken, activeApiTokenRawName) => {
+        const { name } = activeApiToken;
+        if (!!name && name !== activeApiTokenRawName) {
+          return `Your key will be named: ${name}`;
+        }
+      },
+    ],
+  }),
+  listeners: ({ actions, values }) => ({
+    initializeCredentialsData: () => {
+      actions.fetchCredentials();
+      actions.fetchDetails();
+    },
+    fetchCredentials: (page = 1) => {
+      HttpLogic.values.http
+        .get(`/api/app_search/credentials`, {
+          query: {
+            'page[current]': page,
+          },
+        })
+        .then(({ meta, results }) => {
+          actions.setCredentialsData(meta, results);
+        });
+      // TODO handle errors
```

Comment:
I thought about adding Error Handling in this PR, but I wasn't sure exactly how it should work. I also thought it would be easier to write after it's connected to the UI, so I thought about adding this in a subsequent PR.

Reply:
It's actually fairly simple, I think we can do it now if no objections :) It fits in with my (upcoming) requests to convert promise chains to try/catches in any case.

Reply:
You're right, it was fairly straight forward. Thank you for adding.